src/Processors/QueryPlan/Optimizations/mergeFilterIntoJoinCondition.cpp

```diff
@@ -0,0 +1,361 @@
+        return {};
+
+    auto left_stream_allowed_nodes = getInputNodes(filter_dag, left_stream_available_columns);
+    auto right_stream_allowed_nodes = getInputNodes(filter_dag, right_stream_available_columns);
+
+    /// Extract all conjuncts from filter expression
+    auto conjuncts_list = getConjunctsList(predicate);
+
+    JoinConditionParts result;
+    std::unordered_set<const ActionsDAG::Node *> conjuncts_to_replace;
+    ActionsDAG::NodeRawConstPtrs rejected_conjuncts;
+    rejected_conjuncts.reserve(conjuncts_list.size());
+
+    for (const auto * conjunct : conjuncts_list)
+    {
+        bool is_equality = conjunct->type == ActionsDAG::ActionType::FUNCTION && conjunct->function_base->getName() == "equals";
+        if (is_equality)
+        {
+            const auto * lhs = conjunct->children[0];
+            const auto * rhs = conjunct->children[1];
+
+            /// We need to check if arguments are coming from different sides of JOIN
+            auto lhs_side = getExpressionSide(lhs, left_stream_allowed_nodes, right_stream_allowed_nodes);
+            auto rhs_side = getExpressionSide(rhs, left_stream_allowed_nodes, right_stream_allowed_nodes);
+
+            if (lhs_side == ExpressionSide::LEFT && rhs_side == ExpressionSide::RIGHT)
+            {
+                result.emplace_back(createConditionPart(lhs, rhs));
+                conjuncts_to_replace.insert(conjunct);
+                continue;
+            }
+            else if (rhs_side == ExpressionSide::LEFT && lhs_side == ExpressionSide::RIGHT)
+            {
+                result.emplace_back(createConditionPart(rhs, lhs));
+                conjuncts_to_replace.insert(conjunct);
+                continue;
+            }
+        }
+        rejected_conjuncts.push_back(conjunct);
+    }
+
+    bool trivial_filter = rejected_conjuncts.empty();
+    if (!result.empty())
+    {
+        /// There's a non-empty list of extracted condition parts.
+        /// After JOIN step these equalities will always evaluate to true.
+        for (const auto * & output : filter_dag.getOutputs())
+        {
+            auto it = conjuncts_to_replace.find(output);
+            if (it != conjuncts_to_replace.end())
+            {
+                output = &filter_dag.addColumn(ColumnWithTypeAndName(
+                    output->result_type->createColumnConst(1, 1),
+                    output->result_type,
+                    output->result_name));
+            }
+        }
+
+        if (rejected_conjuncts.size() == 1)
+        {
+            filter_dag.addOrReplaceInOutputs(filter_dag.addAlias(*rejected_conjuncts[0], filter_name));
+        }
+        else if (rejected_conjuncts.size() > 1)
+        {
+            FunctionOverloadResolverPtr func_builder_and = std::make_unique<FunctionToOverloadResolverAdaptor>(std::make_shared<FunctionAnd>());
+            filter_dag.addOrReplaceInOutputs(filter_dag.addFunction(func_builder_and, std::move(rejected_conjuncts), filter_name));
```

Comment:
It's also a good idea to verify the return type.
It can change e.g. from Nullable(UInt8) to UInt8